### DISCUSSION TITLE
Is it possible to use local storage to reduce pyodide re-loading time?

### DISCUSSION THREAD

**Discussion by albertas** [pyodide/pyodide#4243] [General]
I wonder would it be possible to store loaded pyodide binary in the browser's local storage? This binary could be retrieved and used when page is reloaded instead of loading pyodide once again from scratch?

**Comment by hoodmane** [comment]
Not in `localStorage` since that only stores strings, but possible in `IndexedDB`. But I think people have looked into this and the results have been disappointing.

Recently I've been investigating taking memory snapshots which could be stored in `IndexedDB` and does lead to excellent startup time. In node, we can get ~40 to 60 ms startups including importing numpy with snapshots and precompiled wasm. Obviously in the browser loading time is going to be worse because of wasm compilation but it could probably be cut down a whole lot on reload.

**Reply by albertas** [reply]
Sounds really promising, do we have a ticket for this POC, where the finds are being shared? I would love to try the experiment myself.

It seems that caching compiled Webassembly modules via IndexedDB might be no longer supported, quote from https://udn.realityripple.com/docs/WebAssembly/Caching_modules

> Warning: Experimental [WebAssembly.Module](https://udn.realityripple.com/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Module) IndexedDB serialization support is being removed from browsers; see [bug 1469395](https://bugzilla.mozilla.org/show_bug.cgi?id=1469395) and [this spec issue](https://github.com/WebAssembly/spec/issues/821).

**Reply by hoodmane** [reply]
No, I will try to make one soon. Hopefully the browser performance doesn't end up being too disappointing. I've been investigating this for use in cloudflare workers where conditions are much better.

I have working code for the memory snapshots here:
https://github.com/hoodmane/workerd-python-experiments/blob/master/src/python.mjs#L172-L177

Here is a hello world example from the master branch of that repo:
https://py-hello-world.runtime-playground.workers.dev/
Here's fastapi demo deployed from the fastapi branch:
https://py-fastapi-demo.runtime-playground.workers.dev/

You can see that the first page load takes close to 4 seconds but subsequent requests take more like 100 to 200 ms.
